### PR TITLE
ForEach reconciler

### PR DESCRIPTION
**Pull request by scothis**
The ForEach reconciler can iterator over a slice calling a nested reconciler for each item. The current item is stashed on the context to be retrieved with a CursorStasher.

Nested iterators are allowed so long as the type of the items being iterated over have unique names. We may enhance the generated stash key in the future to relax this requirement, so users should not assume the specific key to be stable between releases.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -535,6 +536,45 @@ func TenTimesReconciler() *reconcilers.SubReconciler[*buildv1alpha1.Function] {
...
 
+#### ForEach
+
+A [`ForEach`](https://pkg.go.dev/reconciler.io/runtime/reconcilers#ForEach) calls the reconciler for each item returned from Items. The current cursor can be retrieved with [`CursorStasher`](https://pkg.go.dev/reconciler.io/runtime/reconcilers#CursorStasher).
+
+Nested iteration is allowed so long as the types being iterated over contain unique names. Otherwise the stash keys will collide. For testing the nested reconciler outside the scope of the loop, use the stasher's Key method to lookup the StashKey, do not expect the StashKey to be stable between releases.
+
+**Example:**
+
+A ForEach can be used interact with each volume mount on a pod.
```
Comment (mamachanko) [inline README.md:547]:
```suggestion
A ForEach can be used to interact with each volume mount on a pod. 
```

File: README.md [hunk 2]
```diff
@@ -535,6 +536,45 @@ func TenTimesReconciler() *reconcilers.SubReconciler[*buildv1alpha1.Function] {
 }
 ```
 
+#### ForEach
+
+A [`ForEach`](https://pkg.go.dev/reconciler.io/runtime/reconcilers#ForEach) calls the reconciler for each item returned from Items. The current cursor can be retrieved with [`CursorStasher`](https://pkg.go.dev/reconciler.io/runtime/reconcilers#CursorStasher).
```
Comment (mamachanko) [inline README.md:541]:
```suggestion
A [`ForEach`](https://pkg.go.dev/reconciler.io/runtime/reconcilers#ForEach) calls the reconciler for each item returned from Items. A cursor marks the item being reconciled. The current cursor can be retrieved with [`CursorStasher`](https://pkg.go.dev/reconciler.io/runtime/reconcilers#CursorStasher).
```

File: README.md [hunk 3]
```diff
@@ -535,6 +536,45 @@ func TenTimesReconciler() *reconcilers.SubReconciler[*buildv1alpha1.Function] {
...
+
+Nested iteration is allowed so long as the types being iterated over contain unique names. Otherwise the stash keys will collide. For testing the nested reconciler outside the scope of the loop, use the stasher's Key method to lookup the StashKey, do not expect the StashKey to be stable between releases.
```
Comment (mamachanko) [inline README.md:543]: **[issue, non-blocking]:** When reading _"use the stasher's Key method to lookup the StashKey"_ I am wondering whether stasher refers to any `Stasher` or the `CursorStasher`. Maybe rewording, showing an example test, or referring to test in the codebase would help.

Comment (mamachanko) [inline README.md:543]: **[suggestion, non-blocking]:** When reading _"so long as the types being iterated over contain unique names"_ I was assuming that the items need to have unique names but it's speaking about types. I struggle to parse what it means for a "type" to "contain a name". The doc string of `ForEach` is a little clearer:

```go
// Multiple ForEach reconcilers are nestable so long as the types being iterated over are unique.
```

So the types must be unique then, I think. That makes sense, because we identify each nested reconciler by type.

Maybe this is better:

> Nested iteration is allowed so long as each reconciler's type is unique.

File: reconcilers/flow.go [hunk 4]
```diff
@@ -281,6 +282,112 @@ func (r *While[T]) Reconcile(ctx context.Context, resource T) (Result, error) {
 	return aggregateResult, nil
 }
 
+// ForEach calls the reconcilers for each item. The current value of the iteration is exposed as a
+// Cursor that is available via the CursorStasher helper. Multiple ForEach reconcilers are nestable
+// so long as the types being iterated over are unique.
+type ForEach[Type client.Object, Item any] struct {
+	// Name used to identify this reconciler.  Defaults to `ForEach`.  Ideally unique, but
```
Comment (mamachanko) [inline reconcilers/flow.go:289]:
```suggestion
	// Name used to identify this reconciler.  Defaults to `ForEach`. Ideally unique, but
```

File: reconcilers/flow.go [hunk 5]
```diff
@@ -281,6 +282,112 @@ func (r *While[T]) Reconcile(ctx context.Context, resource T) (Result, error) {
...
+type Cursor[I any] struct {
+	// Index of the current iteration
+	Index int
+	// Length of the current iteration, or -1 if not known and/or unbounded
+	Length int
+	// Item of the current iteration
+	Item I
+}
+
+// CursorStasher creates a Stasher for an Cursor of the generic type
```
Comment (mamachanko) [inline reconcilers/flow.go:383]:
```suggestion
// CursorStasher creates a Stasher for a Cursor of the generic type
```

File: reconcilers/flow_test.go [hunk 6]
```diff
@@ -269,6 +271,132 @@ func TestWhile(t *testing.T) {
...
+		"terminate iteration on error": {
+			Resource: resource.
+				SpecDie(func(d *dies.TestResourceSpecDie) {
+					d.TemplateDie(func(d *diecorev1.PodTemplateSpecDie) {
+						d.SpecDie(func(d *diecorev1.PodSpecDie) {
+							d.ContainerDie("hello", func(d *diecorev1.ContainerDie) {
+								d.Image("world")
+							})
+							d.ContainerDie("die", func(d *diecorev1.ContainerDie) {
+								d.Image("die")
```
Comment (mamachanko) [inline reconcilers/flow_test.go:339]: **[issue, absolutely non-blocking]:** overloading the term "die" with "exit early" is fine, but makes the reader squint

File: reconcilers/flow_test.go [hunk 7]
```diff
@@ -269,6 +271,132 @@ func TestWhile(t *testing.T) {
...
+
+	resource := dies.TestResourceBlank.
+		MetadataDie(func(d *diemetav1.ObjectMetaDie) {
+			d.Namespace(testNamespace)
+			d.Name(testName)
+		}).
+		SpecDie(func(d *dies.TestResourceSpecDie) {
+			d.Fields(map[string]string{})
+		})
+
```
Comment (mamachanko) [inline reconcilers/flow_test.go:289]: **[thought, off-topic]:** In terms of reading order _this_ is the moment where I expect to see the reconciler-under-test. I keep scrolling back and forth between the test cases and the reconciler below. It's absolutely fine. Declaring the controller here would work, but it's not something we do conventionally in our test suite. All good 😄

File: reconcilers/reconcilers.go [hunk 8]
```diff
@@ -174,7 +174,6 @@ func typeName(i interface{}) string {
 	}
 
 	t := reflect.TypeOf(i)
-	// TODO do we need this?
```
Comment (mamachanko) [inline reconcilers/reconcilers.go:177]: it seems we don't?! 😆

**Review by mamachanko — APPROVED** [review]
lgmt! left a few thoughts.

File: README.md (hunk 3 above)
Comment (scothis) [reply README.md:543]:
```suggestion
Nested iteration is allowed so long as the types being iterated over contain unique names. Otherwise the stash keys will collide. For testing the nested reconciler outside the scope of the loop, use the CursorStasher's Key method to lookup the StashKey, do not expect the StashKey to be stable between releases.
```

Comment (scothis) [reply README.md:543]: Types have names in go. For example, the type `corev1.Pod` has the name `"Pod"`. Because the stash key is based on a string, we need to coerce the type to a string. In general this will be transparent to users, but if they have nested iteration of the same type or different types that have the same name, then a collision will occur.

I'll take a point to try to clean this up sooner rather than later.

File: reconcilers/flow_test.go (hunk 6 above)
Comment (scothis) [reply reconcilers/flow_test.go:339]: the other meaning of "die" :)

File: reconcilers/reconcilers.go (hunk 8 above)
Comment (scothis) [reply reconcilers/reconcilers.go:177]: we do now

File: reconcilers/flow_test.go (hunk 7 above)
Comment (scothis) [reply reconcilers/flow_test.go:289]: there are places were we do define a unique reconciler for each test case. Since there are only a couple test cases I think this is cleaner, but if those cases needed to diverge more then it would be worth localizing the reconciler definition.
